### PR TITLE
Refine validator interface

### PR DESCRIPTION
**Pull request by kopiczko**

### CODE REVIEW

File: pkg/validation/spec.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	// It returns error matched by IsValidation if the App CR is invalid.
+	//
+	// It returns error matched by IsAppConfigMapNotFound if the ConfigMap
+	// containing values to template is not found in the cluster. Usually
+	// it should be generated by app-operator after a short delay.
+	//
+	// It returns error matched by IsKubeConfigNotFound if the Secret
+	// containing kubeconfig is not generated yet. Usually it should be
+	// generated by app-operator after a short delay.
+	ValidateApp(ctx context.Context, app v1alpha1.App) error
```
Comment (kopiczko) [inline pkg/validation/spec.go:21]: When I was looking at the code it looks to me that only errors matched by `IsValidation` should be matched in app-admission-controller. Otherwise we can get into deadlocks.

Comment (kopiczko) [reply pkg/validation/spec.go:21]: To give more context: I'm working on the admissionkit and in my book validations here should look like:

```go
	err = v.appValidator.ValidateApp(ctx, app)
	if validation.IsAppConfigMapNotFound(err) {
		// Fall trough.
	} else if validation.IsKubeConfigNotFound(err) {
		// Fall trough.
	} else if validation.IsValidationError(err) {
		return validator.Rejectf("validation error: %s", err), nil
	} else if err != nil {
		return nil, microerror.Mask(err)
	}

	return nil, nil // This means accept.
```

File: pkg/validation/spec.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+)
+
+type Interface interface {
+	// ValidateApp checks if given App CR is valid. It returns nil if it is.
+	//
+	// It returns error matched by IsValidation if the App CR is invalid.
+	//
+	// It returns error matched by IsAppConfigMapNotFound if the ConfigMap
+	// containing values to template is not found in the cluster. Usually
+	// it should be generated by app-operator after a short delay.
```
Comment (tomahawk28) [inline pkg/validation/spec.go:16]: From what I know, app-operator doesn't create configmaps belong to the app CRs (e.g. cluster configmap) . there are t other components but in most cases it's cluster-operator.

```suggestion
	// it should be generated by cluster-operator after a short delay.
```

File: pkg/validation/spec.go (hunk 1 above)
Comment (tomahawk28) [reply pkg/validation/spec.go:21]: I don't understand this point, At which point we are ignoring `KubeConfigNotFound` & `AppConfigMapNotFound`? 
I'm asking this since I still see it's only ignore `IsAppDependencyNotReady`
https://github.com/giantswarm/app-admission-controller/blob/3e935256da6026f8a55e8f56d467f4903b94a3ba/pkg/app/validate_app.go#L109-L116

Comment (kopiczko) [reply pkg/validation/spec.go:21]: `IsAppDependencyNotReady` is already gone.

The comment above says what I think it should be. What you linked is what is there.

AIUI values ConfigMap and kubeconfig Secret are created by app-operator. If they are missing then validation fails. This sounds wrong to me.

Comment (tomahawk28) [reply pkg/validation/spec.go:21]: > AIUI values ConfigMap and kubeconfig Secret are created by app-operator/cluster-operator. If they are missing then validation fails. This sounds wrong to me. 

But this is basically what `app-admission-controller` does right now, right?

File: pkg/validation/spec.go (hunk 2 above)
Comment (rossf7) [reply pkg/validation/spec.go:16]: Yes the cluster values configmap and kubeconfig secret are generated by cluster-operator not app-operator.

File: pkg/validation/spec.go (hunk 1 above)
Comment (rossf7) [reply pkg/validation/spec.go:21]: @kopiczko I'm OK if you want to refactor the code for introducing admissionkit. But why do we need to change the validation logic? 

I think these changes may be based on some incorrect assumptions. These resources are managed by cluster-operator not app-operator.. So there is no deadlock but there can be a delay on cluster creation while the cert used by app-operator is issued. 

We need to match on these errors in the app resource that is used by cluster-operator. https://github.com/giantswarm/resource/blob/827f6e7426dced38aab540f9188f6783caae1250/appresource/create.go#L28-L35

Comment (kopiczko) [reply pkg/validation/spec.go:21]: @rossf7 I might miss many details here. So basically values CM and kubeconfig Secret **must** be created before App CR is created?

Comment (tomahawk28) [reply pkg/validation/spec.go:21]: values CM and kubeconfig Secrets are created by cluster-operator, and `mutation` package in app-admission-controller does assign these to app CRs (if those kubeconfig in spec are empty) . Even if we pass validation app CRs which linked to the missing kubeconfig, it would result in failures from app-operator. so I don't think we necessarily pass validations in these cases.

Comment (rossf7) [reply pkg/validation/spec.go:21]: The problem is during the creation of new workload clusters. cluster-operator needs to do multiple things and there is a delay while the app-operator cert is issued. We use this cert to create the kubeconfig.

But this delay is pretty short and unless something is broken (like vault being down) its always shorter than the cluster booting. So cluster creation is not affected. By the time we need to bootstrap chart-operator everything is ready.

In cluster-operator we handle these errors and log them but don't error. This took a while to get working but I'm happy with it and I don't want to have to re-test everything.

TLDR;

I'm fine with refactoring to add admissionkit. We now have 3 admission controllers and keeping them in sync is hard.

But I'd prefer there not to be changes to the validation and defaulting logic. Apart from what you need for config-controller.

Comment (kopiczko) [reply pkg/validation/spec.go:21]: OK, thanks for clarifying.

This is still important to distinguish between validation error and some unexpected error from the webhook standpoint. Then the validation would look like below. Note that returning `validator.Rejectf(...), nil` and `nil, err` would result in a different HTTP response code.

```go
	err = v.appValidator.ValidateApp(ctx, app)
	if validation.IsValidationError(err) {
		// REJECT response (400 bad request).
		//	- Does not result in error level log.
		// 	- Does increment error metric.
		return validator.Rejectf("validation error: %s", err), nil
	} else if err != nil {
		 // ERROR response (500 server error).
		//	- Error log printed.
		// 	- Error metric incremented.
		return nil, microerror.Mask(err)
	}

	return nil, nil // ACCEPT response.
```

Comment (rossf7) [reply pkg/validation/spec.go:21]: Yes agree we need to distinguish between validation and unexpected errors. Using validator.Rejectf and separate HTTP response codes is nice. 👍 

As long as we return separate errors and can still match them like we do here then I'm happy.
https://github.com/giantswarm/app/blob/543654b04f278eb445a99b9e6e5c879af78ca39c/pkg/validation/error.go#L63

**Comment by kopiczko** [comment]
Thanks for your time reviewing this. As I already said in another PR I'm switching my focus to kubebuilder so abandoning this.
